sdk/eventhub/event-hubs/src/pumpManager.ts

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { EventHubClient } from "./eventHubClient";
+import { PartitionContext } from "./partitionContext";
+import { EventPosition } from "./eventPosition";
+import { PartitionProcessor, EventProcessorOptions, CloseReason } from "./eventProcessor";
+import { PartitionPump } from "./partitionPump";
+import * as log from "./log";
+
+/**
+ * The PumpManager handles the creation and removal of PartitionPumps.
+ * It also starts a PartitionPump when it is created, and stops a
+ * PartitionPump when it is removed.
+ * @ignore
+ */
+export class PumpManager {
+  private readonly _eventProcessorName: string;
+  private readonly _options: EventProcessorOptions;
+  private _partitionIdToPumps: {
+    [partitionId: string]: PartitionPump | undefined;
+  } = {};
+
+  /**
+   * @ignore
+   */
+  constructor(eventProcessorName: string, eventProcessorOptions: EventProcessorOptions = {}) {
+    this._eventProcessorName = eventProcessorName;
+    this._options = eventProcessorOptions;
+  }
+
+  /**
+   * Returns a list of partitionIds that are actively receiving messages.
+   * @ignore
+   */
+  public receivingFromPartitions(): string[] {
+    return Object.keys(this._partitionIdToPumps).filter((id) => {
+      const pump = this._partitionIdToPumps[id];
+      return Boolean(pump && pump.isReceiving);
+    });
+  }
+
+  /**
+   * Creates and starts a PartitionPump.
+   * @param eventHubClient The EventHubClient to forward to the PartitionPump.
+   * @param partitionContext The PartitionContext to forward to the PartitionPump.
+   * @param initialEventPosition The EventPosition to forward to the PartitionPump.
+   * @param partitionProcessor The PartitionProcessor to forward to the PartitionPump.
+   * @param abortSignal Used to cancel pump creation.
+   * @ignore
+   */
+  public async createPump(
+    eventHubClient: EventHubClient,
+    partitionContext: PartitionContext,
+    initialEventPosition: EventPosition,
+    partitionProcessor: PartitionProcessor
+  ): Promise<void> {
+    const partitionId = partitionContext.partitionId;
+    // attempt to get an existing pump
+    const existingPump = this._partitionIdToPumps[partitionId];
+    if (existingPump) {
+      if (existingPump.isReceiving) {
+        log.pumpManager(
+          `[${this._eventProcessorName}] [${partitionId}] The existing pump is running.`
+        );
+        return;
+      }
+      log.pumpManager(
+        `[${this._eventProcessorName}] [${partitionId}] The existing pump is not running.`
+      );
+      await this.removePump(partitionId, CloseReason.OwnershipLost);
+    }
+
+    log.pumpManager(`[${this._eventProcessorName}] [${partitionId}] Creating a new pump.`);
+
+    const pump = new PartitionPump(eventHubClient, partitionContext, partitionProcessor, {
+      ...this._options,
+      initialEventPosition
+    });
+
+    try {
+      await pump.start();
```

Comment:
If I have 3 partitions then it will call my `initialize()` method 3 times?

Reply:
Yes, there is one PartitionProcessor per partition :)